File: packages/python/spaces/starfish_spaces/members.py
"""Space membership — invite-based (member cap) and link-based (open access).

MEMBER join: the owner records the invitee in the roster, mints a space-scoped
member cap, and adds the invitee to the space-wide keyring (if it exists) so
they can decrypt ``enc`` content. The invitee stores a ``{kind:'member'}`` entry.

LINK join: the owner mints an ephemeral Ed/KEM keypair whose *private* key ships
inside a URL-fragment token, adds the ephemeral userId to the roster so the server
grants ``space:member``, and mints a member cap scoped to that ephemeral subject.
Any bearer of the link stores a ``{kind:'link'}`` entry.

REVOCATION: ``remove_space_member`` removes the userId from the server roster so
the server stops granting ``space:member`` to new requests.  For ``enc`` spaces,
call :func:`revoke_space_access` instead — it also rotates the space keyring.
"""

from __future__ import annotations

import json
from typing import TYPE_CHECKING, Any, Callable, Optional, TypedDict

from starfish_protocol.encoding import decode_link_fragment, encode_link_fragment

from starfish_spaces.account_seal import seal_to_self, unseal_from_self
from starfish_spaces.client import (
    add_space_keyring_recipient,
    ensure_space_keyring_recipient,
)
from starfish_spaces.invite_helpers import (
    CapSubject,
    assert_cap_for_me,
    cap_nonce,
    ephemeral_subject_async,
    evict_keyring_member,
    mint_cap,
    parse_join_request,
)
from starfish_spaces.keyed_store import create_composed_store
from starfish_spaces.layout import RECIPIENT_LABEL_LEN
from starfish_spaces.registry import (
    add_joined_space_with_cap,
    add_joined_space_with_link_access,
    add_space_member,
    build_space,
    read_spaces,
    remove_space_member,
    update_spaces_doc,
)
from starfish_spaces.request_verify import sign_kem_sig
from starfish_spaces.space_access_store import (
    hydrate_space_access_store,
    local_space_access_entries,
    save_space_access_entry,
)
from starfish_spaces.token_types import JoinRequest, SpaceInviteLinkToken

if TYPE_CHECKING:
    from starfish_spaces.config import SealedBlob
    from starfish_spaces.session import Session


# ── Utilities ─────────────────────────────────────────────────────────────────


def _recipient_for(sub_kem: str, user_id: str) -> dict[str, str]:
    return {"subKem": sub_kem, "userId": user_id, "label": user_id[:RECIPIENT_LABEL_LEN]}


# ── JoinRequest helpers ───────────────────────────────────────────────────────


def make_join_request(session: "Session") -> str:
    """Build a JSON join-request string from the current session's keys."""
    kem_sig = sign_kem_sig(session.keys["kemPub"], session.keys["edPriv"])
    req: JoinRequest = {
        "edPub": session.keys["edPub"],
        "kemPub": session.keys["kemPub"],
        "userId": session.user_id,
        "kemSig": kem_sig,
    }
    return json.dumps(req)


# ── Space invite store (nonces for full eviction) ─────────────────────────────


class StoredSpaceInvite(TypedDict):
    """Retained invite data (owner-side, keyed by ``{spaceId}:{userId}``)."""

    edPub: str
    kemPub: str
    cap: dict[str, Any]  # {nonce, exp}


_space_invite_store = create_composed_store(
    compose_key=lambda space_id, user_id: f"{space_id}:{user_id}"
)


def save_space_invite_entry(space_id: str, user_id: str, entry: StoredSpaceInvite) -> None:
    """Save an invite entry for future revocation."""
    _space_invite_store.for_(space_id, user_id).set(entry)


def get_space_invite_entry(space_id: str, user_id: str) -> Optional[StoredSpaceInvite]:
    """Retrieve a stored invite entry. Returns ``None`` when absent."""
    return _space_invite_store.for_(space_id, user_id).get()


def clear_space_invite_store() -> None:
    """Clear all entries (e.g. on sign-out)."""
    _space_invite_store.store.clear_all()


def serialize_space_invite_store() -> str:
    """Snapshot the store for persistence across reloads."""
    return _space_invite_store.store.serialize()


def hydrate_space_invite_store(raw: str) -> None:
    """Restore the store after a reload (additive — does not clear existing)."""
    _space_invite_store.store.hydrate(raw)


# ── Direct invite ─────────────────────────────────────────────────────────────


async def invite_to_space(
    session: "Session",
    space_id: str,
    request_json: str,
    can_write: bool = True,
    space_name: Optional[str] = None,
) -> str:
    """Owner: invite an identity into a space.

    Records them in the roster, mints a space-scoped member cap, and adds them
    to the space-wide keyring if it exists.

    Args:
        session:      The owner's session.
        space_id:     The target space.
        request_json: JSON join-request from the invitee.
        can_write:    Whether the member gets write access.
        space_name:   Override the space name in the invite bundle.

    Returns:
        The invite bundle JSON; pass to the invitee who calls :func:`accept_space_invite`.
    """
    req_dict = json.loads(request_json)
    req = await parse_join_request(req_dict, session)
    await add_space_member(session.account_client, space_id, session.user_id, req["userIdHex"], session)

    await ensure_space_keyring_recipient(
        session.content_client, session.keys, space_id,  # type: ignore[arg-type]
        _recipient_for(req["kemPubHex"], req["userIdHex"]),
        session.layout,
    )

    subject = CapSubject(
        edPubHex=req["edPubHex"], kemPubHex=req["kemPubHex"], userIdHex=req["userIdHex"]
    )
    cap = mint_cap(session, subject, "content", session.layout.space_member_scope(space_id, can_write))
    nonce = cap_nonce(cap)
    if nonce:
        save_space_invite_entry(space_id, req["userIdHex"], {"edPub": req["edPubHex"], "kemPub": req["kemPubHex"], "cap": nonce})

    if space_name is None:
        doc = await read_spaces(session.account_client, session)
        space_name = next((s["name"] for s in doc.spaces if isinstance(s, dict) and s.get("id") == space_id), "Space") if doc.spaces else "Space"

    invite = {"spaceId": space_id, "spaceName": space_name, "cap": cap}
    return json.dumps(invite)


async def accept_space_invite(session: "Session", invite_json: str) -> Any:
    """Invitee: accept a space invite — store the cap and register the space."""
    inv = json.loads(invite_json)
    cap = inv.get("cap")
    space_id = inv.get("spaceId")
    if not cap or not space_id:
        raise ValueError("That is not a valid space invite.")
    assert_cap_for_me(cap, session)
    cap_json = json.dumps(cap)
    space = build_space(space_id, inv.get("spaceName") or "")
    await add_joined_space_with_cap(session.account_client, session, space, cap_json)
    save_space_access_entry(space_id, {"kind": "member", "cap": cap_json})
    return space


# ── Link-based joins ──────────────────────────────────────────────────────────


def encode_space_invite_link(origin: str, token: SpaceInviteLinkToken) -> str:
    """Encode a :class:`SpaceInviteLinkToken` into a URL fragment."""
    return encode_link_fragment(origin, "join", dict(token))


def decode_space_invite_link(fragment: str) -> SpaceInviteLinkToken:
    """Decode a space invite link URL fragment."""
    def validate(tok: Any) -> Optional[dict]:
        if not isinstance(tok, dict):
            return None
        if not tok.get("spaceId") or not tok.get("cap") or not tok.get("key"):
            return None
        return tok

    raw = decode_link_fragment(fragment, validate, "That space invite link is malformed or incomplete.")
    return {  # type: ignore[return-value]
        "v": 1,
        "spaceId": raw["spaceId"],
        "spaceName": raw.get("spaceName") or "Space",
        "cap": raw["cap"],
        "key": raw["key"],
        "kemPriv": raw.get("kemPriv"),
        "kemPub": raw.get("kemPub"),
        "write": bool(raw.get("write")),
    }


async def create_space_invite_link(
    session: "Session",
    space_id: str,
    space_name: str,
    write: bool,
    origin: str,
) -> dict[str, Any]:
    """Owner: create a shareable invite link for a space.

    Returns:
        ``{"token": SpaceInviteLinkToken, "link": str}``.
    """
    subject, ed_priv, kem_priv = await ephemeral_subject_async(session)
    ephemeral_user_id = subject["userIdHex"]

    cap = mint_cap(session, subject, "content", session.layout.space_member_scope(space_id, write))
    nonce = cap_nonce(cap)
    if nonce:
        save_space_invite_entry(space_id, ephemeral_user_id, {"edPub": subject["edPubHex"], "kemPub": subject["kemPubHex"], "cap": nonce})

    await add_space_member(session.account_client, space_id, session.user_id, ephemeral_user_id, session)

    await ensure_space_keyring_recipient(
        session.content_client, session.keys, space_id,  # type: ignore[arg-type]
        _recipient_for(subject["kemPubHex"], ephemeral_user_id),
        session.layout,
    )

    token: SpaceInviteLinkToken = {  # type: ignore[assignment]
        "v": 1,
        "spaceId": space_id,
        "spaceName": space_name,
        "cap": cap,
        "key": ed_priv,
        "kemPriv": kem_priv,
        "kemPub": subject["kemPubHex"],
        "write": write,
    }
    return {"token": token, "link": encode_space_invite_link(origin, token)}


async def join_space_by_link(session: "Session", token: SpaceInviteLinkToken) -> Any:
    """Any user: join a space by redeeming an invite link token."""
    access_payload = {
        "cap": token["cap"],
        "key": token["key"],
        "kemPriv": token.get("kemPriv"),
        "kemPub": token.get("kemPub"),
        "write": token.get("write", False),
    }
    sealed = seal_to_self(session, json.dumps(access_payload))
    space = build_space(token["spaceId"], token.get("spaceName") or "")
    await add_joined_space_with_link_access(session.account_client, session, space, sealed)
    save_space_access_entry(token["spaceId"], {
        "kind": "link",
        "cap": token["cap"],
        "key": token["key"],
        "kemPriv": token.get("kemPriv"),
        "kemPub": token.get("kemPub"),
        "write": token.get("write", False),
    })
    return space


# ── Device pairing ────────────────────────────────────────────────────────────


async def add_device_to_space_keyring(
    session: "Session",
    space_id: str,
    device: dict[str, str],
) -> None:
    """Add a device's KEM key as a recipient of a space's keyring.

    Call this after device pairing for each space the paired device should decrypt.
    """
    try:
        await add_space_keyring_recipient(
            session.content_client,
            session.keys,  # type: ignore[arg-type]
            session.layout.keyring_name(space_id),
            _recipient_for(device["kemPub"], device.get("userId") or ""),
        )
    except Exception:
        pass  # best-effort (keyring may not exist for non-enc spaces)


# ── Recover space access ──────────────────────────────────────────────────────


async def recover_space_access(
    session: "Session",
    server: dict[str, Any],
) -> None:
    """Single sign-in hydration: merge server-side caps + sealed link access.

    Args:
        session: The active session.
        server:  ``{"caps": {spaceId: capJson}, "pubAccess": {spaceId: SealedBlob}}``.
    """
    caps: dict[str, str] = server.get("caps") or {}
    pub_access: dict[str, Any] = server.get("pubAccess") or {}

    link_access: dict[str, Any] = {}
    for space_id, sealed_blob in pub_access.items():
        try:
            from starfish_spaces.config import SealedBlob as SealedBlobType
            sealed = sealed_blob if isinstance(sealed_blob, dict) else dict(sealed_blob)
            from starfish_spaces.account_seal import unseal_from_self as _unseal
            raw = _unseal(session, sealed)
            parsed = json.loads(raw)
            if parsed.get("cap") and parsed.get("key"):
                link_access[space_id] = parsed
        except Exception:
            pass

    await hydrate_space_access_store(session.user_id, caps, link_access)

    # Backfill local-only entries to the server.
    local = local_space_access_entries()  # list of (space_id, entry) tuples
    missing_caps = {id_: e for id_, e in local if e.get("kind") == "member" and id_ not in caps}
    missing_links = {id_: e for id_, e in local if e.get("kind") == "link" and id_ not in pub_access}

    if not missing_caps and not missing_links:
        return

    try:
        new_caps = {id_: e["cap"] for id_, e in missing_caps.items()}
        new_pub: dict[str, Any] = {}
        for id_, e in missing_links.items():
            payload = {"cap": e["cap"], "key": e["key"], "kemPriv": e.get("kemPriv"), "kemPub": e.get("kemPub"), "write": e.get("write", False)}
            sealed = seal_to_self(session, json.dumps(payload))
            new_pub[id_] = sealed

        await update_spaces_doc(session.account_client, session, lambda cur: {
            "spaces": cur["spaces"],
            "caps": {**cur["caps"], **new_caps},
            "pubAccess": {**cur["pubAccess"], **new_pub},
        })
    except Exception:
        pass


# ── Revoke space access ───────────────────────────────────────────────────────


async def revoke_space_access(
    session: "Session",
    space_id: str,
    user_id: str,
    opts: dict[str, Any],
) -> dict[str, Any]:
    """Fully evict a space member — rotates keyring AND submits revocation.

    ``opts`` must contain ``"generation": int`` and optionally
    ``"priorRevoked"`` and ``"submitRevocation"`` callback.
    """
    invite = get_space_invite_entry(space_id, user_id)
    if not invite:
        raise ValueError(
            f"revoke_space_access: no stored invite for {user_id!r} on space {space_id!r}"
            " — call save_space_invite_entry or use invite_to_space / create_space_invite_link."
        )

    member_entry = {
        "sub": invite["edPub"],
        "nonce": invite["cap"].get("nonce"),
        "exp": invite["cap"].get("exp"),
        "subKem": invite["kemPub"],
    }

    await evict_keyring_member(
        session.content_client,
        session,
        session.layout.keyring_name(space_id),
        member_entry,
        opts["generation"],
        opts.get("priorRevoked"),
        opts.get("submitRevocation"),
    )

    await remove_space_member(session.account_client, space_id, user_id, session)
    return {"revoked": True}


__all__ = [
    "StoredSpaceInvite",
    "make_join_request",
    "save_space_invite_entry",
    "get_space_invite_entry",
    "clear_space_invite_store",
    "serialize_space_invite_store",
    "hydrate_space_invite_store",
    "invite_to_space",
    "accept_space_invite",
    "encode_space_invite_link",
    "decode_space_invite_link",
    "create_space_invite_link",
    "join_space_by_link",
    "add_device_to_space_keyring",
    "recover_space_access",
    "revoke_space_access",
]

File: packages/ts/spaces/src/members.ts
/**
 * Space membership — invite-based (member cap) and link-based (open access).
 *
 * MEMBER join: the owner records the invitee in the roster, mints a space-scoped
 * member cap, and adds the invitee to the space-wide keyring (if it exists) so they
 * can decrypt `enc` content. The invitee stores a `{kind:'member'}` entry.
 *
 * LINK join: the owner mints an ephemeral Ed/KEM keypair whose *private* key ships
 * inside a URL-fragment token, adds the ephemeral userId to the roster so the server
 * grants `space:member`, and mints a member cap scoped to that ephemeral subject.
 * Any bearer of the link stores a `{kind:'link'}` entry.
 *
 * REVOCATION (roster-only): `removeSpaceMember` removes the userId from the server
 * roster so the server stops granting `space:member` to new requests. This alone is
 * sufficient for non-encrypted spaces. For `enc` spaces call `revokeSpaceAccess`
 * instead — it rotates the space keyring (forward secrecy) AND submits a signed
 * RevocationList so the server immediately rejects the evicted member's cap.
 *
 * DEVICE PAIRING: after pairing, call `addDeviceToSpaceKeyring(session, spaceId, device)`
 * for each space the paired device should decrypt. ONE keyring per space encrypts all
 * `enc` nodes; adding the device once unlocks the whole space's E2EE content.
 */
import type { RevocationEntry, RevocationList } from "@drakkar.software/starfish-protocol"
import { encodeLinkFragment, decodeLinkFragment } from "@drakkar.software/starfish-protocol"

import type { Space, SealedBlob, CapMap, PubAccessMap } from "./config.js"
import type { Session } from "./session.js"
import {
  hydrateSpaceAccessStore,
  localSpaceAccessEntries,
  saveSpaceAccessEntry,
} from "./space-access-store.js"
import type { LinkAccessPayload } from "./space-access-store.js"
import {
  assertCapForMe,
  assertCapNotExpired,
  capNonce,
  ephemeralSubject,
  evictKeyringMember,
  mintCap,
  parseJoinRequest,
} from "./invite-helpers.js"
import { signKemSig } from "./request-verify.js"
import { addSpaceKeyringRecipient, ensureSpaceKeyringRecipient, isKeyringMissing } from "./client.js"
import {
  addJoinedSpaceWithCap,
  addJoinedSpaceWithLinkAccess,
  addSpaceMember,
  buildSpace,
  readSpaces,
  updateSpacesDoc,
  removeSpaceMember,
} from "./registry.js"
import { sealToSelf, unsealFromSelf } from "./account-seal.js"
import { createComposedStore } from "./keyed-store.js"
import type { JoinRequest, SpaceInviteLinkToken } from "./token-types.js"

// ── recipientFor helper ────────────────────────────────────────────────────────

import { RECIPIENT_LABEL_LEN } from "./layout.js"

function recipientFor(subKem: string, userId: string) {
  return { subKem, userId, label: userId.slice(0, RECIPIENT_LABEL_LEN) }
}

// ── JoinRequest export ────────────────────────────────────────────────────────

export type { JoinRequest }

export function makeJoinRequest(session: Session): string {
  const kemSig = signKemSig(session.keys)
  const req: JoinRequest = {
    edPub: session.keys.edPub,
    kemPub: session.keys.kemPub,
    userId: session.userId,
    kemSig,
  }
  return JSON.stringify(req)
}

// ── Space invite store (nonces for full eviction) ─────────────────────────────

export interface StoredSpaceInvite {
  edPub: string
  kemPub: string
  /** Retained cap nonce + expiry for the space member cap (`spaceMemberScope`). */
  cap: { nonce: string; exp: number }
}

// Keyed `${spaceId}:${userId}` → invite.
const _spaceInviteStore = createComposedStore<StoredSpaceInvite, [string, string]>(
  (spaceId, userId) => `${spaceId}:${userId}`,
)

const _siRaw = _spaceInviteStore.store

/** Save an invite entry for future revocation. */
export const saveSpaceInviteEntry = (spaceId: string, userId: string, entry: StoredSpaceInvite): void =>
  _spaceInviteStore.for(spaceId, userId).set(entry)
/** Retrieve a stored invite entry. Returns null when absent. */
export const getSpaceInviteEntry = (spaceId: string, userId: string): StoredSpaceInvite | null =>
  _spaceInviteStore.for(spaceId, userId).get() ?? null
/** Clear all entries (e.g. on sign-out). */
export const clearSpaceInviteStore = (): void => {
  const raw = _siRaw.serialize()
  if (raw === "{}") return
  // Bulk-clear by re-hydrating a blank store snapshot.
  const keys = Object.keys(JSON.parse(raw) as Record<string, unknown>)
  for (const k of keys) _siRaw.clear(k)
}
/** Snapshot the store for persistence across reloads. */
export const serializeSpaceInviteStore = (): string => _siRaw.serialize()
/** Restore the store after a reload (additive — does not clear existing entries). */
export const hydrateSpaceInviteStore = (raw: string): void => _siRaw.hydrate(raw)

interface SpaceInvite {
  spaceId: string
  spaceName: string
  cap: unknown
}

/**
 * Owner: invite an identity into a space. Records them in the roster, mints a
 * space-scoped member cap, and adds them to the space-wide keyring if it exists.
 * Returns the invite bundle JSON.
 */
export async function inviteToSpace(
  session: Session,
  spaceId: string,
  requestJson: string,
  canWrite = true,
  spaceName?: string,
): Promise<string> {
  const req = await parseJoinRequest(requestJson, "That is not a valid join request", session)
  await addSpaceMember(session.accountClient, spaceId, session.userId, req.userId, session)

  // Ensure the space-wide keyring exists, then add the invitee as a recipient.
  await ensureSpaceKeyringRecipient(session, spaceId, recipientFor(req.kemPub, req.userId))

  const cap = await mintCap(
    session,
    { edPubHex: req.edPub, kemPubHex: req.kemPub, userIdHex: req.userId },
    "content",
    session.layout.spaceMemberScope(spaceId, canWrite),
  )
  const nonce = capNonce(cap)
  if (nonce) saveSpaceInviteEntry(spaceId, req.userId, { edPub: req.edPub, kemPub: req.kemPub, cap: nonce })

  let name = spaceName?.trim()
  if (!name) {
    const { spaces } = await readSpaces(session.accountClient, session)
    name = spaces.find((s) => s.id === spaceId)?.name ?? "Space"
  }
  const invite: SpaceInvite = { spaceId, spaceName: name, cap }
  return JSON.stringify(invite)
}

/**
 * Invitee: accept a space invite — store the cap and register the space.
 * Returns the joined space.
 */
export async function acceptSpaceInvite(session: Session, inviteJson: string): Promise<Space> {
  const inv = JSON.parse(inviteJson) as Partial<SpaceInvite>
  const cap = inv.cap as { kind?: string; sub?: string } | undefined
  if (!cap || !inv.spaceId) throw new Error("That is not a valid space invite.")
  assertCapForMe(cap, session.keys.edPub, "That is not a valid space invite.", "This invite was issued for a different identity.")
  const spaceId = inv.spaceId
  const capJson = JSON.stringify(cap)
  const space = buildSpace(spaceId, inv.spaceName ?? "")
  await addJoinedSpaceWithCap(session.accountClient, session, space, capJson)
  saveSpaceAccessEntry(spaceId, { kind: "member", cap: capJson })
  return space
}

// ── Link-based joins ──────────────────────────────────────────────────────────

export type { SpaceInviteLinkToken }

export function encodeSpaceInviteLink(origin: string, token: SpaceInviteLinkToken): string {
  return encodeLinkFragment(origin, "join", token)
}

export function decodeSpaceInviteLink(fragment: string): SpaceInviteLinkToken {
  const raw = decodeLinkFragment<{ spaceId: string; cap: unknown; key: string } & Partial<SpaceInviteLinkToken>>(
    fragment,
    (tok: unknown) => {
      const t = tok as Record<string, unknown>
      return !!t && typeof t.spaceId === "string" && !!t.cap && typeof t.key === "string"
        ? (t as { spaceId: string; cap: unknown; key: string } & Partial<SpaceInviteLinkToken>)
        : null
    },
    "That space invite link is malformed or incomplete.",
  )
  return {
    v: 1,
    spaceId: raw.spaceId,
    spaceName: raw.spaceName ?? "Space",
    cap: raw.cap,
    key: raw.key,
    kemPriv: raw.kemPriv,
    kemPub: raw.kemPub,
    write: !!raw.write,
  }
}

/**
 * Owner: create a shareable invite link for a PUBLIC space.
 *
 * The link is a BEARER token — anyone who holds it can join. Bound the exposure
 * with `opts.ttlSec` / `opts.expiresAt` (the cap's server-enforced `exp`; default
 * 30 days when omitted). The returned `inviteUserId` is the ephemeral member's
 * userId — pass it to `revokeSpaceAccess(session, spaceId, inviteUserId, …)` to
 * kill this one link without affecting other members or links.
 *
 * Bearer links are inherently multi-use: there is no client-only single-use, as
 * the secret lives entirely in the URL fragment and the server counts no
 * redemptions. Use a short TTL and/or revoke after the intended join.
 */
export async function createSpaceInviteLink(
  session: Session,
  spaceId: string,
  spaceName: string,
  write: boolean,
  origin: string,
  opts: { ttlSec?: number; nbf?: number; expiresAt?: number } = {},
): Promise<{ token: SpaceInviteLinkToken; link: string; inviteUserId: string }> {
  const { ek, userId: ephemeralUserId, subject } = await ephemeralSubject(session)
  const capOpts = { ttlSec: opts.ttlSec, nbf: opts.nbf, expiresAt: opts.expiresAt }
  const cap = await mintCap(session, subject, "content", session.layout.spaceMemberScope(spaceId, write), capOpts)
  const nonce = capNonce(cap)
  if (nonce) saveSpaceInviteEntry(spaceId, ephemeralUserId, { edPub: ek.edPub, kemPub: ek.kemPub, cap: nonce })
  // Add the ephemeral userId to the roster
  await addSpaceMember(session.accountClient, spaceId, session.userId, ephemeralUserId, session)

  // Ensure the keyring exists, then add the ephemeral KEM so link-bearers can decrypt enc content.
  await ensureSpaceKeyringRecipient(session, spaceId, recipientFor(ek.kemPub, ephemeralUserId))

  const token: SpaceInviteLinkToken = {
    v: 1, spaceId, spaceName, cap, key: ek.edPriv, kemPriv: ek.kemPriv, kemPub: ek.kemPub, write,
  }
  return { token, link: encodeSpaceInviteLink(origin, token), inviteUserId: ephemeralUserId }
}

/**
 * Any user: join a space by redeeming an invite link token.
 */
export async function joinSpaceByLink(session: Session, token: SpaceInviteLinkToken): Promise<Space> {
  assertCapNotExpired(token.cap, "That space invite link is no longer usable")
  const space = buildSpace(token.spaceId, token.spaceName)
  const accessPayload = { cap: token.cap, key: token.key, kemPriv: token.kemPriv, kemPub: token.kemPub, write: token.write }
  const sealed = await sealToSelf(session, JSON.stringify(accessPayload))
  await addJoinedSpaceWithLinkAccess(session.accountClient, session, space, sealed)
  saveSpaceAccessEntry(token.spaceId, {
    kind: "link", cap: token.cap, key: token.key, kemPriv: token.kemPriv, kemPub: token.kemPub, write: token.write,
  })
  return space
}

/**
 * Add a device's KEM key as a recipient of a space's keyring.
 * Call this after device pairing.
 */
export async function addDeviceToSpaceKeyring(
  session: Session,
  spaceId: string,
  device: { kemPub: string; edPub: string; userId: string },
): Promise<void> {
  try {
    await addSpaceKeyringRecipient(session, spaceId, recipientFor(device.kemPub, device.userId))
  } catch (err) {
    if (!isKeyringMissing(err)) throw err
  }
}

/**
 * Single sign-in hydration: merges server-side caps and sealed link access
 * into the unified space-access store. Call once on sign-in / account switch.
 */
export async function recoverSpaceAccess(
  session: Session,
  server: { caps: Record<string, string>; pubAccess: Record<string, SealedBlob> },
): Promise<void> {
  // Unseal link access blobs
  const linkAccess: Record<string, LinkAccessPayload> = {}
  for (const [spaceId, sealed] of Object.entries(server.pubAccess)) {
    try {
      const raw = await unsealFromSelf(session, sealed)
      const parsed = JSON.parse(raw) as LinkAccessPayload
      if (parsed.cap && parsed.key) linkAccess[spaceId] = parsed
    } catch (e) {
      console.error("[starfish-spaces] recoverSpaceAccess: failed to unseal", spaceId, (e instanceof Error ? e.message : String(e)))
    }
  }

  await hydrateSpaceAccessStore(session.userId, server.caps, linkAccess)

  // Backfill local-only entries to the server
  const local = localSpaceAccessEntries()
  const missingMemberCaps = Object.entries(local)
    .filter(([id, e]) => e.kind === "member" && !(id in server.caps))
  const missingLinks = Object.entries(local)
    .filter(([id, e]) => e.kind === "link" && !(id in server.pubAccess))

  if (missingMemberCaps.length === 0 && missingLinks.length === 0) return

  try {
    const newCaps: CapMap = {}
    for (const [id, e] of missingMemberCaps) if (e.kind === "member") newCaps[id] = e.cap

    const newPubAccess: PubAccessMap = {}
    for (const [id, e] of missingLinks) {
      if (e.kind === "link") {
        newPubAccess[id] = await sealToSelf(
          session,
          JSON.stringify({ cap: e.cap, key: e.key, kemPriv: e.kemPriv, kemPub: e.kemPub, write: e.write }),
        )
      }
    }

    await updateSpacesDoc(session.accountClient, session, (cur) => ({
      spaces: cur.spaces,
      caps: { ...cur.caps, ...newCaps },
      pubAccess: { ...cur.pubAccess, ...newPubAccess },
    }))
  } catch (e) {
    console.error("[starfish-spaces] recoverSpaceAccess: backfill failed", (e instanceof Error ? e.message : String(e)))
  }
}

// ── Space-tier full eviction ──────────────────────────────────────────────────

/**
 * Fully evict a space member — rotates keyring AND submits revocation.
 */
export async function revokeSpaceAccess(
  session: Session,
  spaceId: string,
  userId: string,
  opts: {
    generation: number
    priorRevoked?: RevocationEntry[]
    submitRevocation: (list: RevocationList) => Promise<void>
  },
): Promise<{ revoked: boolean }> {
  const invite = getSpaceInviteEntry(spaceId, userId)
  if (!invite) {
    throw new Error(
      `revokeSpaceAccess: no stored invite for ${userId} on space ${spaceId} — call saveSpaceInviteEntry or use inviteToSpace / createSpaceInviteLink (which auto-store the entry)`,
    )
  }

  await evictKeyringMember(
    session.contentClient,
    session,
    session.layout.keyringName(spaceId),
    { sub: invite.edPub, nonce: invite.cap.nonce, exp: invite.cap.exp, subKem: invite.kemPub },
    { generation: opts.generation, priorRevoked: opts.priorRevoked, submitRevocation: opts.submitRevocation },
  )

  await removeSpaceMember(session.accountClient, spaceId, userId, session)

  return { revoked: true }
}
